### ISSUE TITLE
Include body hash in Detached JWS

### ISSUE THREAD

**Issue by aaronpk** [ietf-wg-gnap/gnap-core-protocol#201]
As I was working on a client implementation, it occurred to me that it would be a lot simpler and less error prone to include a hash of the request body in the JWS rather than doing the current "compute-the-JWS-then-remove-the-body" method.

This would remove the `b64` and `crit` fields from the JWS header, and would add a new sentence describing how to calculate the body hash, dropping the dependency on RFC7797:

> The payload of the JWS object is the SHA-256 hash of the serialized body of the request

This significantly cleans up the client code, since previously the client would have to generate the JWS, the pull out the header and signature components and reassemble them with an empty body, creating a string like "XXXXX..XXXXX". Now it can use a JWS library without any additional hacks on top.

**Comment by adeinega** [comment]
Such JWS header wouldn't qualify RFC7797 in this case then.

**Comment by aaronpk** [comment]
how so?

**Comment by adeinega** [comment]
RFC 7797 requires you to include "b64" in the "crit" header parameter and set the "b64" header parameter to `false`.

Lots of existing JWT parsers obey the "crit" parameter. As you mention, you may remove these parameters. Although, it will result in an inability to parse such JWTs by parsers that already support RFC 7797.

**Comment by adeinega** [comment]
Some of the existing parsers provide a convenient way to get "XXXXX..XXXXX" without any hacks on top. For example, take a look at method `getDetachedContentCompactSerialization(...)` in https://bitbucket.org/b_c/jose4j/wiki/JWS%20Examples#markdown-header-using-the-rfc-7797-jws-unencoded-payload-option.

**Comment by aaronpk** [comment]
Not relying on the additional rules specified in RFC7797 seems like a benefit, less specs to read and less edge cases to handle. Removing a dependency on another spec is a benefit.

**Comment by jricher** [comment]
Right, these wouldn't be RFC7797 values anymore, they would just be plain JWS objects with a specially calculated payload. We still will have the "XXXX..XXXX" format for payload-less requests (requests without the body, such as GET and DELETE).

**Comment by jricher** [comment]
I think I'm in favor of this idea, personally. It still fulfills the requirement of covering the message body with the signature method, but it does so in a way that's perhaps more native to most JOSE developers. It's also more conducive to the kinds of deployment models discussed in #196

**Comment by adeinega** [comment]
I hear you. I would like to add & emphasize that there are two parts to this story. The first part where we create JWTs and accordingly, there is the second part where we verify them. The beauty is that there are lots of libraries that support RFC 7797. The rest of the decently implemented libraries [will clearly indicate that they can't verify such](https://tools.ietf.org/html/rfc7515#page-14) JWTs because they don't know how to proceed with b64 in the crit header parameter and it's really great IMO. It reduces the chances of errors. Now, let’s imagine we omitted _b64_ from the crit header parameter and started to put the SHA-256 hash in the payload of the JWS object. In this case, the verification process should fail by all existing libraries, right? If you decide to put something like “gnap-detached” in the crit header parameter then you will also need to change the code of a JWT library so this library will be able to process it. I am guessing that isn't always possible, rational, or just easy to do.

**Comment by jricher** [comment]
The body of this new method would be a base64 encoded hash value, so we actually don't :want: the `b64` header at all and we wouldn't be doing RFC7797 here. 

One thing that's different from what many libraries may expect is that the body doesn't decode into JSON -- but this was always a JWS, not a JWT. Just about every JWS library I've used, these days, allows for an arbitrary payload. Even https://mkjose.org/ allows for signing an arbitrary string for the payload, but it will base64 encode it first. I think this new approach benefits from not requiring any `crit` parameter, enforcement of which I've found hit-or-miss at best in the JOSE community.
